core/trino-main/src/main/java/io/trino/execution/HeapSizeParser.java

```diff
@@ -0,0 +1,62 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.execution;
+
+import com.google.common.annotations.VisibleForTesting;
+import io.airlift.units.DataSize;
+
+import java.util.function.Supplier;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static java.lang.Double.parseDouble;
+import static java.util.Objects.requireNonNull;
+
+public class HeapSizeParser
+{
+    private static final String RELATIVE_SUFFIX = "%";
+    private static final Supplier<Long> AVAILABLE_HEAP_MEMORY = Runtime.getRuntime()::maxMemory;
```

Comment:
Making it a supplier instead of static constant may be problematic.
We have a code comment in DeltaLakeConfig
```
// Runtime.getRuntime().maxMemory() is not 100% stable and may return slightly different value over JVM lifetime. We use
// constant so default configuration for cache size is stable.
```

Reply:
This is stable as the configuration is resolved once during startup. Since this is part of the `trino-main` it can't be used in the connector.

Reply:
I'll memoize the value so it's computed once. Once Stable values are introduced in Java we will be able to use it instead (which is a nicer thing as they are constant folded)

Reply:
Memoized + documentation added